util/approx_math/approx_math.H

```diff
@@ -0,0 +1,159 @@
+#ifndef APPROX_MATH_H
+#define APPROX_MATH_H
+
+#include <AMReX_REAL.H>
+#include <microphysics_math.H>
+#include <limits>
+
+using namespace amrex::literals;
+
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+amrex::Real fast_atan_1(const amrex::Real x) {
+    ///
+    /// This calculates atan within [-1, 1] range.
+    ///
+
+    ///
+    /// Approximation of atan for x in [-1, 1]
+    /// Max absolute error for this is ~0.0015 rad
+    /// See Ref:
+    /// Efficient Approximations for the Arctangent Function by Rajan 2006
+    ///
+
+    // constexpr amrex::Real PI_4 = GCEM_PI / 4.0_rt;
+
+    // return PI_4*x - x*(std::abs(x) - 1.0_rt) *
+    //     (0.2447_rt + 0.0663_rt*std::abs(x));
+
+
+    ///
+    /// Another approximation of atan for x in [-1, 1]
+    /// Max absolute error for this is ~0.00063 rad
+    /// See Ref:
+    /// https://stackoverflow.com/questions/42537957/fast-accurate-atan-arctan-approximation-algorithm
+    ///
+
+    constexpr amrex::Real A = 0.0776509570923569_rt;
+    constexpr amrex::Real B = -0.287434475393028_rt;
+    constexpr amrex::Real C = GCEM_PI / 4.0_rt - A - B;
+
+    amrex::Real x2 = x*x;
+    return ((A*x2 + B)*x2 + C)*x;
+}
+
+
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+amrex::Real fast_atan(const amrex::Real x) {
+    ///
+    /// Fast atan approximation calculations.
+    ///
+
+    constexpr amrex::Real PI_2 = 0.5_rt * GCEM_PI;
+
+    ///
+    /// If x < 0.113, then using arctan(x) ~ x
+    /// gives better answer than the approximation below.
+    /// And accuracy increase as x << 0.113.
+    /// Also significantly faster.
+    ///
+
+    if (std::abs(x) < 0.113_rt) {
+        return x;
+    }
+
+    // Check for large number, close to infinity.
+    // Error is ~1e-8 rad by not checking actual inf
+
+    if (x > 1.e8_rt) {
+        return PI_2;
+    }
+    if (x < -1.e8_rt) {
+        return -PI_2;
+    }
+
+    // Now calculate Atan(x) using approximations
+
+    if (x > 1.0_rt) {
+        return PI_2 - fast_atan_1(1.0_rt / x);
+    }
+    if (x < -1.0_rt) {
+        return -PI_2 - fast_atan_1(1.0_rt / x);
+    }
+    return fast_atan_1(x);
+}
+
+
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+amrex::Real cawley_exp(const amrex::Real x) {
+    ///
+    /// A fast implementation of exp.
+    /// Roughly 6-times faster than std::exp.
+    /// With reasonable accuracy across all range.
+    ///
+    /// Code is obtained from Ref (2):
+    /// Ref:
+    /// 1) A Fast, Compact Approximation of the Exponential Function
+    ///    by Schraudolph 1999
+    /// 2) On a Fast Compact Approximation of the Exponential Function
+    ///    by Cawley 2000
+    ///
+
+    // Use EXP_A = 1512775 if input is integer.
+    constexpr amrex::Real EXP_A = 1048576.0_rt / GCEM_LOG_2;
+
+    // The Choice of this number is described in Ref 1.
+    constexpr int EXP_C = 60801;
+
+    union {
+        amrex::Real d;
+        struct {
+#ifdef LITTLE_ENDIAN
+            int j, i;
+#else
+            int i, j;
+#endif
```

Comment:
Explicit int sizes are clearer here (also add `#include <cstdint>` at the top)
```suggestion
#ifdef LITTLE_ENDIAN
            std::int32_t j, i;
#else
            std::int32_t i, j;
#endif
```

Reply:
done